torch/jit/_script.py

```diff
@@ -879,8 +877,18 @@ def _script_pdt(obj, optimize=None, _frames_up=0, _rcb=None, example_inputs: Opt
     if monkeytype_trace:
         monkeytype_config = JitTypeTraceConfig(type_trace_db)
         with monkeytype_trace(monkeytype_config):
-            for example_input in example_inputs:  # type: ignore[union-attr]
-                obj(*example_input)
+            for example_input in example_inputs:    # type: ignore[union-attr]
+                # If the obj is an nn.Module or a class, then each method is
+                # executed with the arguments provided in the example inputs.
+                # example inputs here will be of tuple(class.method, (arguments))
+                # This is used to infer type annotations for those methods
+                # which are not called directly under the hood of monkeytype.
+                if isinstance(obj, torch.nn.Module) or inspect.isclass(obj):
+                    for module, example_input in example_inputs:  # type: ignore[union-attr]
```

Comment:
Does this mean user needs to enumerate all submodules in `example_inputs`? I think this is a huge regression in usability. We must find a better solution.

I don't think this should be needed either since running top level module with provided example inputs should trigger execution of submodules too. Thus submodules would be profiled naturally. What am I missing?

Reply:
The user doesn't have to enumerate all the submodules. Only those which we were not invoked by monkeytype explicitly. Like for eg.,

class ExampleModel(nn.Module):
         def __init__(self):
               super.__init__(
        
        @torch.jit.export 
        def fn(self, a, b):
              ....
        @torch.jit.export
        def other_fn(self, z)
             ....
model = ExampleModel()
s = torch.jit._script_pdt(model, example_inputs=[(model.fn, (10, 20, ) , (3.9, 4.9, ),)])
```
So here, the user can choose to not pass `other_fn` and can only selectively annotate `fn`.

Reply:
I see. That sounds more reasonable. Could we allow both input format for `example_inputs` thought?
One is `Optional[List[Tuple]]`, the other is `Optional[Dict[callable, List[Tuple]]]`.

Reply:
So by annotating `example_inputs: Union[Optional[List[Tuple]], Optional[Dict[Callable, List[Tuple]]]] = None`, mypy throws a bunch of errors:

Like in `_script.py` in the else part: 
```
else:
    for example_input in example_inputs:
        obj(*example_input)
# mypy error:  Incompatible types in assignment (expression has type "Union[Tuple[Any, ...], Callable[..., Any]]", variable has type "List[Tuple[Any, ...]]")  [assignment]
```
And for the tests mypy complain the following?
```
inp = [(20, ), (2.7, ), (False, )]
scripted_pdt_model = torch.jit._script_pdt(wrapped_pdt_model, example_inputs={wrapped_pdt_model: inp, })

# mypy error: Dict entry 0 has incompatible type "NestedModulePDTWrapper": "List[Tuple[float]]"; expected "Callable[..., Any]": "List[Tuple[Any, ...]]"  [dict-item]
```
Here mypy identifies the type of `inp` as `List[Tuple[float]]` although it's of type `List[Tuple[Any]]`

Is it ok to suppress these errors?

Reply:
All fixed now.